pkg/gpio/gpio_linux.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package gpio provides functions for interacting with GPIO pins via the
+// GPIO Sysfs Interface for Userspace.
+package gpio
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+const gpioPath = "/sys/class/gpio"
+
+// Value represents the value of a gpio pin
+type Value bool
+
+// Gpio pin values can either be low (0) or high (1)
+const (
+	Low  Value = false
+	High Value = true
+)
+
+func (v Value) Dir() string {
+	if v == Low {
+		return "low"
+	}
+	return "high"
+}
+
+func (v Value) String() string {
+	if v == Low {
+		return "0"
+	}
+	return "1"
+}
+
+func ValueOfInt(val int) (Value, error) {
+	switch val {
+	case 0:
+		return Low, nil
+	case 1:
+		return High, nil
+	default:
+		return Low, fmt.Errorf("Invalid pin value")
+	}
+}
+
+// SetGpioOutputValue configures the gpio as an output pin with the given value.
+// Value should either be 0 (low) or 1 (high).
+func SetGpioOutputValue(pin int, val Value) error {
+	dir := val.Dir()
+	path := filepath.Join(gpioPath, "gpio", strconv.Itoa(pin), "direction")
+	outFile, err := os.OpenFile(path, os.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("failed to open %s: %v", path, err)
+	}
+	defer outFile.Close()
+	if _, err := outFile.WriteString(dir); err != nil {
+		return fmt.Errorf("failed to set gpio %d to %s: %v", pin, dir, err)
+	}
+	return nil
+}
+
+// ReadGpioValue returns the value of the given gpio pin. If the read was
+// unsuccessful, it returns a value of Low and the associated error.
+func ReadGpioValue(pin int) (Value, error) {
+	path := filepath.Join(gpioPath, "gpio", strconv.Itoa(pin), "value")
+	buf, err := ioutil.ReadFile(path)
+	if err != nil {
+		return Low, fmt.Errorf("failed to read value of gpio %d: %v", pin, err)
+	}
+	val, err := strconv.Atoi(strings.TrimSuffix(string(buf), "\n"))
```

Comment:
This is fine but I end up wondering if you need all the mechanism of atoi and trimsuffix
switch string(buf) {
case "0\n": return Low, nil
case "1\n": return High, nil
}
return Low, fmt.Errorf("failed to read value of gpio %d (%s): %v", pin, string(buf),  err)
}

I figure you might want to return the value of buf as it makes figuring out what is wrong a bit easier.

Reply:
Oh I like this much better!